File: mtui/mcp/deny.py
"""REPL-only commands that the MCP server must not expose as tools.

Each entry cannot meaningfully run outside an interactive terminal session
and is filtered out when ``mtui.mcp.tools`` synthesises tools from
``mtui.commands.Command.registry``.

Per-entry rationale:

- ``quit``, ``exit``, ``EOF``: call ``sys.exit`` and would tear down the
  MCP server process along with the client connection.
- ``edit``: spawns ``$EDITOR`` on ``metadata.path``; replaced by the
  ``testreport_read`` / ``testreport_patch`` / ``testreport_write`` MCP
  tools that operate on the loaded testreport file directly.
- ``shell``: opens an interactive root PTY on a refhost and needs a TTY
  attached to the client, which MCP transports do not provide.
- ``help``: prints argparser help text to stdout; the MCP protocol
  already advertises tool descriptions to clients.
- ``terms``: launches local terminal-emulator scripts (``term.<name>.sh``)
  that spawn ``xterm``/``konsole``/etc. on the operator's ``$DISPLAY``.
- ``switch``: interactive navigation that moves the session's active-template
  pointer — REPL-only state with no client-addressable equivalent under MCP
  (tools select a template per call via the ``template`` parameter, and
  ``list_templates`` remains a read-only listing).

``unload`` is **not** denied: it takes an explicit RRID, mutates only the
loaded set (closing that template's host connections), needs no TTY, and does
not call ``sys.exit``, so it is exposed as an MCP tool. It is the addressable
counterpart to ``load_template`` for dropping one template from a session.
"""

REPL_ONLY: frozenset[str] = frozenset(
    {"quit", "exit", "EOF", "edit", "shell", "help", "terms", "switch"}
)

File: crates/mtui-mcp/src/deny.rs
//! Commands the MCP server must not expose as tools.
//!
//! Each entry either cannot meaningfully run outside an interactive terminal
//! session, or is replaced by a richer hand-written tool; [`crate::tools`]
//! filters them out when synthesising tools from the command
//! [`mtui_core::Registry`]. (`lrun` needs no denying: the command was removed
//! from mtui entirely.)
//!
//! The deny surface is **not** re-declared here: it is the single
//! [`mtui_core::MCP_DENYLIST`], which sits beside `register_all` and is
//! consistency-checked against the registry there. This module is the thin
//! MCP-side accessor.
//!
//! - `quit`, `exit`, `EOF`: exit the process, tearing the server down with it.
//! - `edit`: spawns `$EDITOR` on the controlling TTY; the testreport tools
//!   operate on the loaded report file directly instead.
//! - `shell`: an interactive root PTY needs a TTY the MCP transports lack.
//! - `help`: the MCP protocol already advertises tool descriptions.
//! - `terms`: launches terminal-emulator scripts on the operator's `$DISPLAY`.
//! - `switch`: REPL-only active-template pointer; tools select a template per
//!   call via the `template` parameter.
//! - `get`, `put`: their synthesized forms exchange **server-local paths** a
//!   remote `--transport http` client cannot reach. The hand-written tools in
//!   [`crate::transfer_tools`] carry the content in-band under the same names
//!   instead (#434) — the `edit` → testreport-tools precedent, with name reuse
//!   made collision-free by this very deny.
//!
//! `unload` is deliberately **not** denied: it names an explicit RRID, mutates
//! only the loaded set, needs no TTY, and does not exit — the addressable
//! counterpart to `load_template`.

pub use mtui_core::MCP_DENYLIST;

/// Whether `name` must not become an MCP tool.
#[must_use]
pub(crate) fn is_denied(name: &str) -> bool {
    MCP_DENYLIST.contains(&name)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn blocked_commands_are_denied() {
        for name in [
            "quit", "exit", "EOF", "edit", "shell", "help", "terms", "switch", "get", "put",
        ] {
            assert!(is_denied(name), "{name} must be denied");
        }
    }

    #[test]
    fn exposed_commands_are_not_denied() {
        for name in ["run", "update", "whoami", "unload", "config", "list_hosts"] {
            assert!(!is_denied(name), "{name} must not be denied");
        }
    }
}
